tests/operators/test_python_operator.py

```diff
@@ -242,6 +249,510 @@ def test_echo_env_variables(self):
+            ref=object1
+        )
+
+        task = PythonOperator(
+            task_id='python_operator',
+            python_callable=build_recording_function(recorded_calls),
+            provide_context=True,
+            templates_dict={
+                'a_templated_object': object2
+            },
+            dag=self.dag)
+
+        self.dag.create_dagrun(
+            run_id='manual__' + DEFAULT_DATE.isoformat(),
+            execution_date=DEFAULT_DATE,
+            start_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+        task.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
+
+        self.assertEqual(1, len(recorded_calls))
+        self.assertTrue('templates_dict' in recorded_calls[0].kwargs)
+        self.assertDictEqual(
+            recorded_calls[0].kwargs['templates_dict'],
+            {
+                'a_templated_object': ClassWithCustomAttributes(
+                    templated_string_attr="object 2 on dag {}".format(self.dag.dag_id),
+                    ref=ClassWithCustomAttributes(
+                        templated_string_attr="object 1 on dag {}".format(self.dag.dag_id),
+                        ref=None
+                    )
+                )
+            }
+        )
+
+    def test_nested_template_fields_can_be_defined_at_the_class_level(self):
+        recorded_calls = []
+        ClassWithCustomAttributes.template_fields = ['templated_string_attr']
+
+        object1 = ClassWithCustomAttributes(
+            templated_string_attr="object 1 on dag {{dag.dag_id}}"
+        )
+        object2 = ClassWithCustomAttributes(
+            templated_string_attr="object 2 on dag {{dag.dag_id}}"
+        )
+
+        task = PythonOperator(
+            task_id='python_operator',
+            python_callable=build_recording_function(recorded_calls),
+            provide_context=True,
+            templates_dict={
+                'an_object': object1,
+                'another_object': object2
+            },
+            dag=self.dag)
+
+        self.dag.create_dagrun(
+            run_id='manual__' + DEFAULT_DATE.isoformat(),
+            execution_date=DEFAULT_DATE,
+            start_date=DEFAULT_DATE,
+            state=State.RUNNING
+        )
+        task.run(start_date=DEFAULT_DATE, end_date=DEFAULT_DATE)
+
+        self.assertEqual(1, len(recorded_calls))
+        self.assertTrue('templates_dict' in recorded_calls[0].kwargs)
+        self.assertDictEqual(
+            recorded_calls[0].kwargs['templates_dict'],
+            {
+                'an_object': ClassWithCustomAttributes(
+                    templated_string_attr="object 1 on dag {}".format(self.dag.dag_id)
+                ),
+                'another_object': ClassWithCustomAttributes(
+                    templated_string_attr="object 2 on dag {}".format(self.dag.dag_id)
+                )
+            }
+        )
+
+    def test_nested_template_fields_can_be_defined_at_the_instance_level(self):
```

Comment:
I don't think we need this test. I understand what you've done, but this is more testing python classes.

Basically I feel that this test case is just testing this: 

```python
class X:
   x = 1

i = X()
assert i.x == 1
```

and 

```python
class X:
    pass

i = X()
i.x = 1
assert i.x == 1
```